### PR TITLE
Some monadic rewrite rules

### PR DESCRIPTION
**Pull request by michaelmcinerney**

### CODE REVIEW

**Review by lsf37 — APPROVED** [review]
Looks good from my side.

File: lib/MonadicRewrite.thy [hunk 1]
```diff
@@ -32,6 +32,10 @@ lemma monadic_rewrite_refl_generic:
 
 lemmas monadic_rewrite_refl = monadic_rewrite_refl_generic[where P=\<top>]
 
+lemma monadic_rewrite_sym:
+  "monadic_rewrite False True P f g \<Longrightarrow> monadic_rewrite False True P g f"
+  by (simp add: monadic_rewrite_def)
```
Comment (Xaphiosis) [inline lib/MonadicRewrite.thy:37]: This is really high up, given the `F E` flags are explicitly set, so it's a sym rule only for that case. Does it still hold if the `F` flag is left generic?

File: lib/MonadicRewrite.thy [hunk 2]
```diff
@@ -178,6 +182,12 @@ lemma monadic_rewrite_named_bindE:
   apply simp
   done
 
+lemma monadic_rewrite_drop_return:
+  "monadic_rewrite F E P f g \<Longrightarrow> monadic_rewrite F E P (do y \<leftarrow> return r; f od) g"
```
Comment (Xaphiosis) [inline lib/MonadicRewrite.thy:186]: suggest `return x` or something that hints at a value... `r` is usually a relation/set.

Comment (lsf37) [reply lib/MonadicRewrite.thy:186]: also happier with `return x`, but I'd like to point out that a set should be `R` not `r` (a relation could be either).

File: lib/MonadicRewrite.thy (hunk 1 above)
Comment (michaelmcinerney) [reply lib/MonadicRewrite.thy:37]: I'm happy to move this. Where would be best, do you think?

Unfortunately, I couldn't see any generalisation of this for either the `F` or `E` flag

Comment (Xaphiosis) [reply lib/MonadicRewrite.thy:37]: Well, if this is the only way symmetry works, then leave it where it is. If we have doubts, or there's something we're paving over, maybe dump it at the end of this section after `monadic_rewrite_weaken2`?

Comment (michaelmcinerney) [reply lib/MonadicRewrite.thy:37]: I've taken another look and I'm sure that we can't generalise this any further, so I think I might leave it where it is

**Comment by michaelmcinerney** [comment]
Looking at the output of the `lemmas` command more closely, it wasn't what I wanted. But what I did want is very straightforward and is now schematic in the flags, too

**Comment by Xaphiosis** [comment]
> Looking at the output of the `lemmas` command more closely, it wasn't what I wanted. But what I did want is very straightforward and is now schematic in the flags, too

Good.
That would mean the `sym` rule is now unused? Do we still need/want it?

**Comment by michaelmcinerney** [comment]
> Good. That would mean the `sym` rule is now unused? Do we still need/want it?

That's right, it's now currently unused in `master`. It's used in several places for MCS, so it might be useful. I'd be happy to remove it, but maybe it would be good to have this file in sync for both branches

**Comment by Xaphiosis** [comment]
> > Good. That would mean the `sym` rule is now unused? Do we still need/want it?
> 
> That's right, it's now currently unused in `master`. It's used in several places for MCS, so it might be useful. I'd be happy to remove it, but maybe it would be good to have this file in sync for both branches

If it's used on MCS then we can leave it. The question is whether it's used well on MCS, or more like a hammer looking for nails. Probably can do that another time though.

**Pull request merged**